File: backend/app/schemas/candidate_risk.py
"""Pydantic schemas for the Candidate Risk Profile feature.

See `app.models.candidate_risk` and migracja 0066_candidate_risk for the
underlying data model. Mirrors backend enums to allow direct serialization.
"""

from datetime import datetime
from typing import List, Literal, Optional

from pydantic import BaseModel, Field

from app.models.candidate_risk import RiskLevel


RiskCategory = Literal["early", "interview", "post_accept"]


class RiskBreakdown(BaseModel):
    """Liczność wycofań w 24-miesięcznym oknie per kategoria."""

    early: int = Field(0, ge=0, description="Wycofania przed jakimkolwiek interview")
    interview: int = Field(
        0, ge=0, description="Wycofania po wejściu w interview / cv_sent / client_interview"
    )
    post_accept: int = Field(
        0, ge=0, description="Wycofania po akcepcie oferty przez kandydata"
    )


class RiskEvent(BaseModel):
    """Pojedynczy dropout event w historii kandydata."""

    job_id: int
    moved_at: datetime
    category: RiskCategory
    reason: str
    job_title: Optional[str] = None


class CandidateRiskProfileOut(BaseModel):
    """Response payload dla GET /api/candidates/{id}/risk."""

    candidate_id: int
    level: RiskLevel
    score: int = Field(..., ge=0)
    breakdown: RiskBreakdown
    last_updated_at: datetime
    recent_events: List[RiskEvent] = Field(default_factory=list)
    profile_exists: bool = True

    model_config = {"from_attributes": True}

File: frontend/src/types/candidate-risk.ts
/**
 * Candidate Risk Profile types – mirror of `app/schemas/candidate_risk.py`.
 *
 * Backend route: `GET /api/candidates/{id}/risk`
 * Migracja:     0068_candidate_risk
 */

export type RiskLevel = "low" | "medium" | "high";

export type RiskCategory = "early" | "interview" | "post_accept";

export interface RiskBreakdown {
  early: number;
  interview: number;
  post_accept: number;
}

export interface RiskEvent {
  job_id: number;
  moved_at: string;
  category: RiskCategory;
  reason: string;
  job_title?: string | null;
}

export interface CandidateRiskProfile {
  candidate_id: number;
  level: RiskLevel;
  score: number;
  breakdown: RiskBreakdown;
  last_updated_at: string | null;
  recent_events: RiskEvent[];
  profile_exists: boolean;
}

export type CandidateOfferResponse = "pending" | "accepted" | "declined";

/**
 * Stage'e w pipeline gdzie kandydat już zaakceptował ofertę. Wycofanie z tych
 * stage'ów z `candidate_offer_response='declined'` = post_accept dropout (10pt).
 * Trzymamy lustro `POST_ACCEPT_STAGES` z `app/services/candidate_risk.py`.
 */
export const POST_ACCEPT_STAGES = new Set<string>([
  "acceptance",
  "negotiation",
  "onboarding",
]);
